src/Mod/PartDesign/App/FeatureHole.cpp

```diff
@@ -1379,31 +1379,40 @@ double Hole::getThreadProfileAngle()
 
 void Hole::findClosestDesignation()
 {
-    // Intended for thread type changes
-    // finds the closest diameter of the new thread type
     int threadType = ThreadType.getValue();
-    if (threadType == -1) {
+    const int numTypes = static_cast<int>(std::size(threadDescription)); 
+
+    if (threadType < 0 || threadType >= numTypes) {
         throw Base::IndexError(QT_TRANSLATE_NOOP("Exception", "Thread type is invalid"));
     }
-    int closestSize = 0;
+
     double diameter = ThreadDiameter.getValue();
-    if (diameter == 0)
+    if (diameter == 0.0) {
         diameter = Diameter.getValue();
-    double closestDifference = std::numeric_limits<double>::infinity();
-    double difference;
+    }
 
-    for (size_t i = 0; i < threadDescription[threadType].size(); i++) {
-        difference = threadDescription[threadType][i].diameter - diameter;
-        if (difference == 0) {
-            closestSize = i;
-            break;
-        }
-        if (std::abs(difference) < closestDifference) {
-            closestSize = i;
-            closestDifference = std::abs(difference);
+    int oldSizeIndex = ThreadSize.getValue();
+    const auto &options = threadDescription[threadType];
+    double targetPitch = 0.0;
+    if (oldSizeIndex >= 0 && oldSizeIndex < static_cast<int>(options.size())) {
+        targetPitch = options[oldSizeIndex].pitch;
+    }
+
+    // Scan all entries to find the minimal (Δdiameter, Δpitch) Euclidean distance
+    size_t bestIndex = 0;
+    double bestMetric = std::numeric_limits<double>::infinity();
+
+    for (size_t i = 0; i < options.size(); ++i) {
+        double dDiff = options[i].diameter - diameter;
+        double pDiff = options[i].pitch - targetPitch;
+        double metric = std::hypot(dDiff, pDiff);
+        if (metric < bestMetric) {
+            bestMetric = metric;
+            bestIndex = i;
         }
     }
-    ThreadSize.setValue(closestSize);
+
+    ThreadSize.setValue(static_cast<int>(bestIndex));

```

Comment:
How about:
```cpp
auto bestIt = std::ranges::min_element(options, {}, [&](const auto& option) {
    double dDiff = option.diameter - diameter;
    double pDiff = option.pitch - targetPitch;
    return std::hypot(dDiff, pDiff);
});

size_t bestIndex = bestIt == options.end() ? 0 : std::distance(options.begin(), bestIt);
ThreadSize.setValue(static_cast<int>(bestIndex));
```

Reply:
@pierreporte mentioned that it would be good to prioritize diameter first.

By returning a pair instead of the result from `hypot` would do that if `dDiff` is placed in the first slot.
It would look something like:
```c++
auto bestIt = std::ranges::min_element(options, {}, [&](const auto& option) {
    double dDiff = std::abs(option.diameter - diameter);
    double pDiff = std::abs(option.pitch - targetPitch);
    return std::make_pair(dDiff, pDiff);
});

size_t bestIndex = bestIt == options.end() ? 0 : std::distance(options.begin(), bestIt);
ThreadSize.setValue(static_cast<int>(bestIndex));
```